Point cloud Cycles primitive

dedoardo opened the pull request:
This patch implements the `HdPoints` primitive using the new Cycles point cloud. The following features are supported:
- Point styles (SPHERE, DISC, DISC_ORIENTED) with radius/normal per point
- Motion blur through vertex velocity/acceleration (geometry) or through transforms (object)
- Particle color and transparency 

Some things that could be needed:
- Arbitrary shaders. Some of the internal shading attributes (e.g. uv) are currently not set for points, need to investigate which ones would be needed.
- Time varying radius and normals
- Motion blur through time sampled vertex positions (Least priority)

This patch requires:
- [Cycles PR](https://github.com/tangent-opensource/cycles/pull/29)
- [UsdCycles PR](https://github.com/tangent-opensource/usdcycles/pull/2)
- (For houdini) [houdini_cycles PR](https://github.com/tangent-opensource/houdini_cycles/pull/1)

bareya commented:
Just an observation, this code will not compile without `usdCycles` schema support.  I am getting following errors:

```
/home/bareya/tangent/dev/hdcycles_for_houdini/build/hdcycles-prefix/src/hdcycles/plugin/hdCycles/points.cpp:133:61: error: ‘usdCyclesTokens’ was not declared in this scope
                                                             usdCyclesTokens->primvarsCyclesObjectMblur,

```

Maybe we should merge following PR first?
https://github.com/tangent-opensource/hdcycles/pull/122

dedoardo commented:
Good catch, forgot to wrap `_ReadObjectFlags` in #ifdefs. I never compile without `usdCycles` :)  

Waiting for #122 sounds good to me.

bareya commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
+            TF_WARN("Point style primvar exists, but is empty for point cloud %s", id.GetText());
+        } else {
+            TfToken point_style = point_style_.Cast<TfToken>().UncheckedGet<TfToken>();
+            if (point_style == usdCyclesTokens->sphere) {
```
This seems like a perfect opportunity for appending new dirty bits. A good example would be normals in `HdStMesh`: 

```
    enum DirtyBits : HdDirtyBits {
        DirtySmoothNormals  = HdChangeTracker::CustomBitsBegin,
        DirtyFlatNormals    = (DirtySmoothNormals << 1),
        DirtyIndices        = (DirtyFlatNormals   << 1),
        DirtyHullIndices    = (DirtyIndices       << 1),
        DirtyPointsIndices  = (DirtyHullIndices   << 1)
    };
```

`PropagateDirtyBits` sounds like a perfect place to implement behaviour for new dirty bits.

dedoardo replied:
The point style are propagates the dirty points, what is the point of defining this new enum?

bareya commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
+HdDirtyBits
+HdCyclesPoints::GetInitialDirtyBitsMask() const
+{
+    return HdChangeTracker::DirtyPoints | HdChangeTracker::DirtyTransform | HdChangeTracker::DirtyVisibility
```
Missing `DirtyPrimvar`? if we decided to add support for arbitrary attribute

dedoardo replied:
Added

bareya commented on plugin/hdCycles/renderParam.cpp:
```diff
@@ -1737,6 +1741,23 @@ HdCyclesRenderParam::AddMesh(ccl::Mesh* a_mesh)
 }
 
+void
+HdCyclesRenderParam::AddPointCloud(ccl::PointCloud* a_pc)
```
`ccl::PointCloud` is a polymorphic type, do we really need this separation for curve, mesh, and point cloud? They are all sub-types of `ccl::Geometry`.

bareya commented on plugin/hdCycles/renderParam.cpp:
```diff
@@ -1737,6 +1741,23 @@ HdCyclesRenderParam::AddMesh(ccl::Mesh* a_mesh)
+        return;
+    }
+
+    m_pointCloudUpdated = true;
```
I am not sure what does this do. Where is it used? `m_geometryUpdated` isn't enough?

bareya commented on plugin/hdCycles/renderParam.cpp:
```diff
@@ -1838,6 +1859,28 @@ HdCyclesRenderParam::RemoveMesh(ccl::Mesh* a_mesh)
 }
 
+void
+HdCyclesRenderParam::RemovePointCloud(ccl::PointCloud* a_pc)
```
If we get rid of `m_pointCloudUpdated` can we remove this function and make two generic functions: `AddGeometry` and `RemoveGeometry`?

bareya commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
-                param->AddObject(pointObject);
-            }
-        }
+    if (!value_.IsHolding<VtFloatArray>()) {
```
`VtValue` offers `CanCast` and `Cast` methods. Appropriate "Casters" are registered in `HdCyclesAttributeSource` implementation.

dedoardo replied:
Tried to follow the usage in Mesh, let me know if it's ok.

bareya commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
+    ccl::float4* C = attr_C->data_float4();
+    if (interpolation == HdInterpolationConstant) {
+        assert(value.size() == 1);
+        const ccl::float3 v0 = vec3f_to_float3(value[0]);
```
Unnecessary conversion

bareya commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
+
+    if (interpolation == HdInterpolationConstant) {
+        assert(value.size() == 1);
+        const ccl::float3 N0 = vec3f_to_float3(value[0]);
```
Unnecessary conversion.

bareya reviewed (changes requested):
Great stuff Edoardo! I tested your PR and it does render with small code alterations such as `TypeFloat4` :)

Unfortunately, I experienced some problems that point cloud does not appear in Houdini when I click `Restart Render`. First time it does render. Second time it disappears. 

Also, there is a bit of code duplication. If you wish investing time in removing the duplication that would be awesome. 

During implementing support for arbitrary attributes for curves I implemented very generic interface for delayed Attribute creation. It could be easily extended to support `ATTR_STD` creation.

dedoardo commented:
@bareya Yes, I wasn't finished with the changes. Changed the target of the PR.

dedoardo commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
+    ccl::Attribute* attr_C = attributes->find(attrib_name);
+    bool reset_opacity     = false;
+    if (!attr_C) {
+        attr_C        = attributes->add(attrib_name, ccl::TypeDesc::TypeFloat4, ccl::ATTR_ELEMENT_VERTEX);
```
Thanks, changed.

dedoardo commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
-    HdCyclesPDPIMap pdpi;
+    m_cyclesPointCloud = new ccl::PointCloud();
+    assert(m_cyclesPointCloud);
+    m_cyclesPointCloud->point_style = (ccl::PointCloudPointStyle)m_pointStyle;
```
Changed to local variable

dedoardo commented:
Fixes:
- Fixed the issue with Houdini not showing the render(uninitialized var)
- Made the point style variable local
- Use the usd casting stuff
- Add ATTR_STD_GENERATED
- Changed the type to RGBA

Questions:
- If I understand correctly, the attribute source can only be used with attributes with no semantics (custom I guess). I can't find an example of it being used in the code outside the tests, I tried forwarding all the primvars through it but get crashes (reasonably). How do I filter which attributes are custom? 
I was hoping I could use it for any attribute... How does this reduce code duplication?
- Can add dirty bits, but see comment.

dedoardo commented on plugin/hdCycles/points.cpp:
```diff
@@ -94,269 +97,669 @@ HdCyclesPoints::Finalize(HdRenderParam* renderParam)
+        info.second = sceneDelegate->GetPrimvarDescriptors(id, info.first);
+    }
+
+    for (auto& interpolation_description : primvars_desc) {
```
How do I know which attributes are to be forwarded to Cycles?

dedoardo commented:
Actually... no crashes with the attribute source. Is this the correct usage?

dedoardo commented on plugin/hdCycles/points.cpp:
```diff
@@ -730,6 +730,10 @@ HdCyclesPoints::Sync(HdSceneDelegate* sceneDelegate, HdRenderParam* renderParam,
+            } else {
+                HdCyclesPointCloudAttributeSource source {
+                    description.name, HdPrimvarRoleTokens->none, { value }, m_cyclesPointCloud, interpolation
+                };
```
ahh, forgot to `Resolve`...

bareya replied:
You don't need to call `Resolve`. You need to grab an `HdInstance` and create shared pointer of `HdCyclesAttributeSource`. There are aliases for it. Then append that source to HdInstance. The `HdCyclesResourceRegistry` will handle committing resources for you in scene-thread safe manner. Also, it will check if `HdCyclesAttributeSource` is a valid source, if not, it will not commit that resource.

dedoardo commented:
Thanks for the explanation @bareya, it all makes more sense now :) I also found an example in the hair geometry, should have looked more carefully.
Have all the comments been addressed ? 

Some comments/questions:
- Some files coming out of Houdini can be vectors of __int64 (sourceprim for example). This right now triggers an error as the type is not in the hydra enumerator. I need to check the cycles side, but I think we could look to add support for it in the future. 
- Should we check the shader network of the object to see if it needs a specific attribute before adding it to the attributes?
- One thing which is pending are the points dpi. I don't really have a use case for it though. thus added a warning in the code, what do you think?

bareya commented:
Hey @dedoardo looks good and I can't wait to test it! 

I can answer some questions:

> Some files coming out of Houdini can be vectors of __int64 (sourceprim for example). This right now triggers an error as the type is not in the hydra enumerator. I need to check the cycles side, but I think we could look to add support for it in the future.

`VtValue` offers "casting" interface. It's quite easy to register new casting functions I move them [here](https://github.com/tangent-opensource/hdcycles/blob/main/plugin/hdCycles/attributeSource.cpp#L577). It seemed like more appropriate place to keep in `attributeSource.cpp`. For convenience I added helper functions to register new `VtArray<>` conversions. It should be as easy as:

```c++
TryRegisterValArrayCast<float, long long int>();
```

> Should we check the shader network of the object to see if it needs a specific attribute before adding it to the attributes?

I thought about it, but I never implemented that check. I very very vaguely remember from our standups that we had a conversation about need for that check, and I think you mentioned that the attribute need requirement check has no effect until very moment of committing to a device. To be honest I didn't have that much time to go through Cycles' code to actually confirm it, but I will rely on your expertise. Feel free to educate me. Also, If you think we need it to reduce memory consumption, please do it :)
